system/lib/pthread/pthread_join.c

```diff
@@ -9,8 +9,30 @@
 #include <pthread.h>
 
 extern int __pthread_join_js(pthread_t thread, void **retval);
+
 int __pthread_join(pthread_t thread, void **retval) {
+  if (thread->self != thread) {
+    // attempt to join a thread which does not point to a valid thread, or does
+    // not exist anymore.
+    return ESRCH;
+  }
+  if (thread->detach_state == DT_DETACHED) {
```

Comment:
Do we not need to access this atomically? JS seemed to (on `detached`, which I can't seem to find.) It seems to be defined as "volatile", not atomic, in pthread_impl.h

Reply:
Yes, musl is using volatile here for these locations rather than atomics.   This is what they do for all their atomics api in `src/internal/atomic.h`.

I think this pre-check is equivalent to the existing one.  We still need to handle the case where the thread becomes detached *while* are joining it.. and I don't thing the JS code currently does that.. but that is a separate issue I think.

Reply:
I changed this to use `emscripten_atomic_load_u32` to match the existing JS impl.    And I added TODO.

Reply:
having made that change I'm not sure its really necessary.  

musl code internally seems perfectly happy to read this variable without any special atomic functions:

https://github.com/emscripten-core/musl/blob/85e0e3519655220688e757b9d5bfd314923548bd/src/thread/pthread_join.c#L16

and:

https://github.com/emscripten-core/musl/blob/85e0e3519655220688e757b9d5bfd314923548bd/src/thread/pthread_join.c#L35

In these cases I think as long as we don't get tearing then its fine to read out-of-date values (and race with other threads).

see